In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')

import warnings
warnings.filterwarnings('ignore')
import re

# 데이터-속성 매칭
### HappyNewYear님의 공유 코드 사용
* https://dacon.io/competitions/official/235863/codeshare/3877?page=1&dtype=recent

In [2]:
d_code = pd.read_csv('./_data/속성_D_코드.csv', index_col=0).T.to_dict()
h_code = pd.read_csv('./_data/속성_H_코드.csv', index_col=0).T.to_dict()
l_code = pd.read_csv('./_data/속성_L_코드.csv', index_col=0).T.to_dict()

t1 = pd.read_csv('./_data/train.csv',index_col=0)
t2 = pd.read_csv('./_data/test.csv',index_col=0)

In [3]:
from typing import Dict
def add_code(
    df: pd.DataFrame,
    d_code: Dict[int, Dict[str, int]], 
    h_code: Dict[int, Dict[str, int]], 
    l_code: Dict[int, Dict[str, int]],
) -> pd.DataFrame:
    
    # Copy input data
    df = df.copy()   

    # D Code
    df['person_prefer_d_1_n'] = df['person_prefer_d_1'].apply(lambda x: d_code[x]['속성 D 세분류코드'])
    df['person_prefer_d_1_s'] = df['person_prefer_d_1'].apply(lambda x: d_code[x]['속성 D 소분류코드'])
    df['person_prefer_d_1_m'] = df['person_prefer_d_1'].apply(lambda x: d_code[x]['속성 D 중분류코드'])
    df['person_prefer_d_1_l'] = df['person_prefer_d_1'].apply(lambda x: d_code[x]['속성 D 대분류코드'])

    df['person_prefer_d_2_n'] = df['person_prefer_d_2'].apply(lambda x: d_code[x]['속성 D 세분류코드'])
    df['person_prefer_d_2_s'] = df['person_prefer_d_2'].apply(lambda x: d_code[x]['속성 D 소분류코드'])
    df['person_prefer_d_2_m'] = df['person_prefer_d_2'].apply(lambda x: d_code[x]['속성 D 중분류코드'])
    df['person_prefer_d_2_l'] = df['person_prefer_d_2'].apply(lambda x: d_code[x]['속성 D 대분류코드'])

    df['person_prefer_d_3_n'] = df['person_prefer_d_3'].apply(lambda x: d_code[x]['속성 D 세분류코드'])
    df['person_prefer_d_3_s'] = df['person_prefer_d_3'].apply(lambda x: d_code[x]['속성 D 소분류코드'])
    df['person_prefer_d_3_m'] = df['person_prefer_d_3'].apply(lambda x: d_code[x]['속성 D 중분류코드'])
    df['person_prefer_d_3_l'] = df['person_prefer_d_3'].apply(lambda x: d_code[x]['속성 D 대분류코드'])

    df['contents_attribute_d_n'] = df['contents_attribute_d'].apply(lambda x: d_code[x]['속성 D 세분류코드'])
    df['contents_attribute_d_s'] = df['contents_attribute_d'].apply(lambda x: d_code[x]['속성 D 소분류코드'])
    df['contents_attribute_d_m'] = df['contents_attribute_d'].apply(lambda x: d_code[x]['속성 D 중분류코드'])
    df['contents_attribute_d_l'] = df['contents_attribute_d'].apply(lambda x: d_code[x]['속성 D 대분류코드'])

    # H Code
    df['person_prefer_h_1_l'] = df['person_prefer_h_1'].apply(lambda x: h_code[x]['속성 H 대분류코드'])
    df['person_prefer_h_1_m'] = df['person_prefer_h_1'].apply(lambda x: h_code[x]['속성 H 중분류코드'])
    
    df['person_prefer_h_2_l'] = df['person_prefer_h_2'].apply(lambda x: h_code[x]['속성 H 대분류코드'])
    df['person_prefer_h_2_m'] = df['person_prefer_h_2'].apply(lambda x: h_code[x]['속성 H 중분류코드'])
    
    df['person_prefer_h_3_l'] = df['person_prefer_h_3'].apply(lambda x: h_code[x]['속성 H 대분류코드'])
    df['person_prefer_h_3_m'] = df['person_prefer_h_3'].apply(lambda x: h_code[x]['속성 H 중분류코드'])

    df['contents_attribute_h_l'] = df['contents_attribute_h'].apply(lambda x: h_code[x]['속성 H 대분류코드'])
    df['contents_attribute_h_m'] = df['contents_attribute_h'].apply(lambda x: h_code[x]['속성 H 중분류코드'])

    # L Code
    df['contents_attribute_l_n'] = df['contents_attribute_l'].apply(lambda x: l_code[x]['속성 L 세분류코드'])
    df['contents_attribute_l_s'] = df['contents_attribute_l'].apply(lambda x: l_code[x]['속성 L 소분류코드'])
    df['contents_attribute_l_m'] = df['contents_attribute_l'].apply(lambda x: l_code[x]['속성 L 중분류코드'])
    df['contents_attribute_l_l'] = df['contents_attribute_l'].apply(lambda x: l_code[x]['속성 L 대분류코드'])
    
    return df

In [21]:
df_train = add_code(t1, d_code, h_code, l_code)
df_test = add_code(t2, d_code, h_code, l_code)

### 훈련 데이터는 1~11월

In [5]:
# train 월데이터 
sorted(df_train['contents_open_dt'].apply(lambda x: x.split('-')[1]).unique())

['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11']

### 테스트 데이터는 12월의 자료로 구성되어있음

In [6]:
# test 월데이터 
sorted(df_test['contents_open_dt'].apply(lambda x: x.split('-')[1]).unique())

['12']

### 따라서 월 데이터는 의미가 없고 요일과 시간 정보는 사용할 수 있을 것으로 보임.


In [7]:
df_train['contents_open_dt']

id
0         2020-01-17 12:09:36
1         2020-06-18 17:48:52
2         2020-07-08 20:00:10
3         2020-01-13 18:09:34
4         2020-03-09 20:39:22
                 ...         
501946    2020-03-13 12:55:52
501947    2020-01-20 11:51:51
501948    2020-08-05 17:27:24
501949    2020-06-15 09:23:21
501950    2020-05-25 14:34:48
Name: contents_open_dt, Length: 501951, dtype: object

컨텐츠 열람에 따라 그루핑

In [10]:
df_train.target

id
0         1
1         0
2         0
3         0
4         0
         ..
501946    1
501947    1
501948    1
501949    1
501950    1
Name: target, Length: 501951, dtype: int64

In [8]:
df_train

d_l_match_yn  d_m_match_yn  d_s_match_yn  h_l_match_yn  h_m_match_yn  \
id                                                                             
0               True          True          True         False         False   
1              False         False         False          True          True   
2              False         False         False          True         False   
3              False         False         False          True         False   
4               True          True          True         False         False   
...              ...           ...           ...           ...           ...   
501946         False         False         False          True         False   
501947          True          True         False          True         False   
501948          True          True          True          True         False   
501949          True         False         False          True         False   
501950          True          True          True          True         False   

        h_s_match_yn  person_attribute_a  person_attribute_a_1  \
id                                                               
0              False                   1                     4   
1              False                   1                     3   
2              False                   2                     0   
3              False                   2                     0   
4              False                   1                     3   
...              ...                 ...                   ...   
501946         False                   1                     1   
501947         False                   1                     6   
501948         False                   1                     7   
501949         False                   1                     1   
501950         False                   1                     6   

        person_attribute_b  person_prefer_c  ...  person_prefer_h_2_l  \
id                                           ...                        
0                        3                5  ...                   94   
1                        4                1  ...                   94   
2                        3                5  ...                  250   
3                        2                5  ...                  226   
4                        4                5  ...                  208   
...                    ...              ...  ...                  ...   
501946                   2                2  ...                    3   
501947                   2                1  ...                   94   
501948                   4                1  ...                   94   
501949                   2                1  ...                   48   
501950                   3                5  ...                   30   

        person_prefer_h_2_m  person_prefer_h_3_l  person_prefer_h_3_m  \
id                                                                      
0                       398                   58                  368   
1                       400                   94                  399   
2                       528                   48                  366   
3                       495                    1                  315   
4                       481                  208                  480   
...                     ...                  ...                  ...   
501946                  316                   94                  398   
501947                  425                   94                  398   
501948                  413                   94                  422   
501949                  359                  250                  523   
501950                  342                   48                  359   

        contents_attribute_h_l  contents_attribute_h_m  \
id                                                       
0                           94                     422   
1                           94    

In [12]:
X = df_train[['person_attribute_a','person_attribute_a_1', 'person_attribute_b']]

In [15]:
from sklearn.cluster import KMeans
SEED = 42
km = KMeans(n_clusters=3,
           init='k-means++',
           n_init=10,
           max_iter=300,
           tol=1e-04,
           random_state=SEED)
y_km = km.fit_predict(X)

In [16]:
y_km

array([2, 2, 0, ..., 1, 0, 1])

In [22]:
df_train['y_km(k3)']= y_km

In [23]:
df_train[['y_km(k3)','target']]

y_km(k3)  target
id                      
0              2       1
1              2       0
2              0       0
3              0       0
4              2       0
...          ...     ...
501946         0       1
501947         1       1
501948         1       1
501949         0       1
501950         1       1

[501951 rows x 2 columns]

In [24]:
km = KMeans(n_clusters=5,
           init='k-means++',
           n_init=10,
           max_iter=300,
           tol=1e-04,
           random_state=SEED)
y_km = km.fit_predict(X)

In [25]:
df_train['y_km(k5)']= y_km
df_train[['y_km(k5)','target']]

y_km(k5)  target
id                      
0              1       1
1              2       0
2              3       0
3              3       0
4              2       0
...          ...     ...
501946         3       1
501947         4       1
501948         4       1
501949         3       1
501950         4       1

[501951 rows x 2 columns]

# 동일한 속성 매칭 및 순서형 데이터 처리
### 열람일시 데이터로 시간, 요일 정보 추가
### 같은 알파벳은 동일한 속성을 의미. 속성 매칭 여부 정보 추가

In [26]:
from sklearn.preprocessing import LabelEncoder

def match_cols(df, del_columns=False):
    
    del_cols = ['person_attribute_a','person_prefer_f','person_prefer_g', 'contents_open_dt', 'person_rn', 'contents_rn']
    
    
    # d속성과 h속성 일치여부
    p_cols = []
    c_d_cols = []
    c_h_cols = []    
    for i in df.columns:
        if re.findall('person_prefer_._._',i):
            p_cols.append(i)
        if re.findall('contents_attribute_d_.',i):
            c_d_cols.append(i)    
        if re.findall('contents_attribute_h+',i):
            c_h_cols.append(i)
            
    for i in p_cols:
        if i.split('_')[2] == 'd':
            for cols in c_d_cols:
                if i.split('_')[-1] == cols[-1]:
                    df[i[-5:]+'_match_yn'] = (df[i]==df[cols]).astype(int)
        else:
            for cols in c_h_cols:
                if i.split('_')[-1] == cols[-1]:
                    df[i[-5:]+'_match_yn'] = (df[i]==df[cols]).astype(int)
    
    # 회원속성a와 컨텐츠 속성a 일치여부 
#     df['match_a'] = (df.person_attribute_a==df.contents_attribute_a).astype(int)  
    
    # 회원속성 a는 1,2 컨텐츠 속성 a는 1~3까지있음
#     df['contents_attribute_a3'] = (df.contents_attribute_a==3).astype(int) 

    # 선호속성c와 컨텐츠 속성c 일치여부
    df['match_c'] = (df.person_prefer_c==df.contents_attribute_c).astype(int)  
    # 회원선호속성 c는 1~5까지 있는 반면, 컨텐츠 속성c는 1~4만 있음
    # 5 = 무응답이나 기타일것같은.. 없으면 성능 떨어짐
    df['person_prefer_c5'] = (df.person_prefer_c==5).astype(int) 
    
    # 선호속성e와 컨텐츠 속성e 일치여부
    df['match_e'] = (df.person_prefer_e==df.contents_attribute_e).astype(int)
    # person_prefer_e 와 contents_attribute_e차이
    df['differ_e'] = df['person_prefer_e'] - df['contents_attribute_e']
    

    # 회원 속성 조합
    le = LabelEncoder()

    df['person_attribute_a_a1'] = df['person_attribute_a'].astype(str)+df['person_attribute_a_1'].astype(str)
    df['person_attribute_a_a1_b'] = df['person_attribute_a'].astype(str)+df['person_attribute_a_1'].astype(str)+df['person_attribute_b'].astype(str)
    df['person_attribute_a_b'] = df['person_attribute_a'].astype(str)+df['person_attribute_b'].astype(str)
    df['person_attribute_a1_b'] = df['person_attribute_a_1'].astype(str)+df['person_attribute_b'].astype(str)
    
    y = le.fit_transform(df['person_attribute_a_a1'].values)
    df['person_attribute_a_a1'] = y
    y = le.fit_transform(df['person_attribute_a_a1_b'].values)
    df['person_attribute_a_a1_b'] = y
    y = le.fit_transform(df['person_attribute_a_b'].values)
    df['person_attribute_a_b'] = y
    y = le.fit_transform(df['person_attribute_a1_b'].values)
    df['person_attribute_a1_b'] = y
    
    # 컨텐츠 속성 조합
    # j*j_1
    df['contents_attribute_j_j1'] = df['contents_attribute_j'].astype(str)+df['contents_attribute_j_1'].astype(str)
    y = le.fit_transform(df['contents_attribute_j_j1'].values)
    df['contents_attribute_j_j1'] = y
    
#     df['contents_attribute_a_e'] = df['contents_attribute_a'].astype(str)+df['contents_attribute_e'].astype(str)
#     y = le.fit_transform(df['contents_attribute_a_e'].values)
#     df['contents_attribute_a_e'] = y

    df['person_attribute_a1_e'] = df['person_attribute_a_1'].astype(str)+df['contents_attribute_e'].astype(str)
    y = le.fit_transform(df['person_attribute_a1_e'].values)
    df['person_attribute_a1_e'] = y
    
    del_cols += ['contents_attribute_a','contents_attribute_j']
    
    # 순서가 있는 속성
    df = pd.get_dummies(df, columns=['person_attribute_a_1','person_attribute_b','person_prefer_e'])

    #contents_open_dt
    df['contents_open_dt'] = pd.to_datetime(df['contents_open_dt'])
    df['contents_open_hour'] = df['contents_open_dt'].dt.hour
    df['contents_open_day'] = df['contents_open_dt'].dt.dayofweek
    
    if del_columns:
        del_cols += (p_cols + c_d_cols + c_h_cols)
                     
    new_df = df.drop(del_cols, axis=1)
    new_df = new_df.drop(columns=['person_prefer_c','contents_attribute_c'])#
    #person_attribute_a
    return new_df

In [27]:
train = match_cols(df_train)
test = match_cols(df_test)

In [10]:
train.columns

Index(['d_l_match_yn', 'd_m_match_yn', 'd_s_match_yn', 'h_l_match_yn',
       'h_m_match_yn', 'h_s_match_yn', 'person_prefer_d_1',
       'person_prefer_d_2', 'person_prefer_d_3', 'person_prefer_h_1',
       ...
       'person_prefer_e_4', 'person_prefer_e_5', 'person_prefer_e_6',
       'person_prefer_e_7', 'person_prefer_e_8', 'person_prefer_e_9',
       'person_prefer_e_10', 'person_prefer_e_11', 'contents_open_hour',
       'contents_open_day'],
      dtype='object', length=105)

# CatBoostClassifier 사용

In [28]:
from catboost import Pool,CatBoostClassifier
import sklearn 
from sklearn.model_selection import StratifiedKFold , KFold
from sklearn.metrics import f1_score 


In [40]:
y_train = train['target']
X_train = train.drop(['target'],axis=1)
X_test = test

In [43]:
#X_train = X_train.drop('contents_attribute_j_1', axis=1)

In [44]:
# f, ax = plt.subplots(figsize=(30, 30))
# ax = sns.heatmap(pd.concat([X_train,y_train],axis=1).corr(), cmap="YlGnBu", vmax=.3, annot=True, linewidths=.5)

In [30]:
cat_features = X_train.columns[X_train.nunique() > 2].tolist()
cat_features

['person_prefer_d_1',
 'person_prefer_d_2',
 'person_prefer_d_3',
 'person_prefer_h_1',
 'person_prefer_h_2',
 'person_prefer_h_3',
 'contents_attribute_i',
 'contents_attribute_j_1',
 'contents_attribute_l',
 'contents_attribute_d',
 'contents_attribute_m',
 'contents_attribute_e',
 'contents_attribute_h',
 'person_prefer_d_1_n',
 'person_prefer_d_1_s',
 'person_prefer_d_1_m',
 'person_prefer_d_1_l',
 'person_prefer_d_2_n',
 'person_prefer_d_2_s',
 'person_prefer_d_2_m',
 'person_prefer_d_2_l',
 'person_prefer_d_3_n',
 'person_prefer_d_3_s',
 'person_prefer_d_3_m',
 'person_prefer_d_3_l',
 'contents_attribute_d_n',
 'contents_attribute_d_s',
 'contents_attribute_d_m',
 'contents_attribute_d_l',
 'person_prefer_h_1_l',
 'person_prefer_h_1_m',
 'person_prefer_h_2_l',
 'person_prefer_h_2_m',
 'person_prefer_h_3_l',
 'person_prefer_h_3_m',
 'contents_attribute_h_l',
 'contents_attribute_h_m',
 'contents_attribute_l_n',
 'contents_attribute_l_s',
 'contents_attribute_l_m',
 'contents_attri

In [41]:
# k5만 사용
X_train = X_train.drop('y_km(k3)',axis=1)
cat_features = X_train.columns[X_train.nunique() > 2].tolist()
cat_features

['person_prefer_d_1',
 'person_prefer_d_2',
 'person_prefer_d_3',
 'person_prefer_h_1',
 'person_prefer_h_2',
 'person_prefer_h_3',
 'contents_attribute_i',
 'contents_attribute_j_1',
 'contents_attribute_l',
 'contents_attribute_d',
 'contents_attribute_m',
 'contents_attribute_e',
 'contents_attribute_h',
 'person_prefer_d_1_n',
 'person_prefer_d_1_s',
 'person_prefer_d_1_m',
 'person_prefer_d_1_l',
 'person_prefer_d_2_n',
 'person_prefer_d_2_s',
 'person_prefer_d_2_m',
 'person_prefer_d_2_l',
 'person_prefer_d_3_n',
 'person_prefer_d_3_s',
 'person_prefer_d_3_m',
 'person_prefer_d_3_l',
 'contents_attribute_d_n',
 'contents_attribute_d_s',
 'contents_attribute_d_m',
 'contents_attribute_d_l',
 'person_prefer_h_1_l',
 'person_prefer_h_1_m',
 'person_prefer_h_2_l',
 'person_prefer_h_2_m',
 'person_prefer_h_3_l',
 'person_prefer_h_3_m',
 'contents_attribute_h_l',
 'contents_attribute_h_m',
 'contents_attribute_l_n',
 'contents_attribute_l_s',
 'contents_attribute_l_m',
 'contents_attri

In [48]:
cat_features = X_train.columns[X_train.nunique() > 2].tolist()
cat_features = cat_features[:-9]+cat_features[-8:] # l포함
# cat_features = cat_features[:-14]+ ['y_km(k3)']+ cat_features[-8:] #

cat_features

['person_prefer_d_1',
 'person_prefer_d_2',
 'person_prefer_d_3',
 'person_prefer_h_1',
 'person_prefer_h_2',
 'person_prefer_h_3',
 'contents_attribute_i',
 'contents_attribute_j_1',
 'contents_attribute_l',
 'contents_attribute_d',
 'contents_attribute_m',
 'contents_attribute_e',
 'contents_attribute_h',
 'person_prefer_d_1_n',
 'person_prefer_d_1_s',
 'person_prefer_d_1_m',
 'person_prefer_d_1_l',
 'person_prefer_d_2_n',
 'person_prefer_d_2_s',
 'person_prefer_d_2_m',
 'person_prefer_d_2_l',
 'person_prefer_d_3_n',
 'person_prefer_d_3_s',
 'person_prefer_d_3_m',
 'person_prefer_d_3_l',
 'contents_attribute_d_n',
 'contents_attribute_d_s',
 'contents_attribute_d_m',
 'contents_attribute_d_l',
 'person_prefer_h_1_l',
 'person_prefer_h_1_m',
 'person_prefer_h_2_l',
 'person_prefer_h_2_m',
 'person_prefer_h_3_l',
 'person_prefer_h_3_m',
 'contents_attribute_h_l',
 'contents_attribute_h_m',
 'contents_attribute_l_n',
 'contents_attribute_l_s',
 'contents_attribute_l_m',
 'contents_attri

In [49]:
len(cat_features)

50

In [37]:
SEED = 42
is_holdout = False
n_splits = 5
iterations = 3000
patience = 100

cv = KFold(n_splits=n_splits, shuffle=True, random_state=SEED)

In [44]:
scores = []
models = []
for tri, vai in cv.split(X_train):
    print("="*50)
    preds = []

    model = CatBoostClassifier(iterations=iterations,random_state=SEED,task_type="CPU",eval_metric="F1",cat_features=cat_features)#,one_hot_max_size=4
    model.fit(X_train.iloc[tri], y_train[tri], 
            eval_set=[(X_train.iloc[vai], y_train[vai])], 
            early_stopping_rounds=patience ,
            verbose = 100
        )
    
    models.append(model)
    scores.append(model.get_best_score()["validation"]["F1"])
    if is_holdout:
        break 

Learning rate set to 0.086395
0:	learn: 0.6137407	test: 0.6150481	best: 0.6150481 (0)	total: 1.17s	remaining: 58m 34s
100:	learn: 0.6574908	test: 0.6867941	best: 0.6869865 (96)	total: 2m 43s	remaining: 1h 18m 7s
200:	learn: 0.6645301	test: 0.6905215	best: 0.6905215 (200)	total: 5m 33s	remaining: 1h 17m 19s
300:	learn: 0.6688147	test: 0.6916508	best: 0.6919988 (291)	total: 8m 43s	remaining: 1h 18m 11s
400:	learn: 0.6715843	test: 0.6927965	best: 0.6929303 (398)	total: 11m 50s	remaining: 1h 16m 45s
500:	learn: 0.6733001	test: 0.6930398	best: 0.6932696 (492)	total: 15m	remaining: 1h 14m 49s
600:	learn: 0.6749028	test: 0.6932861	best: 0.6934715 (528)	total: 18m 8s	remaining: 1h 12m 25s
700:	learn: 0.6763686	test: 0.6940157	best: 0.6940597 (699)	total: 21m 16s	remaining: 1h 9m 46s
800:	learn: 0.6777907	test: 0.6942825	best: 0.6943430 (771)	total: 24m 31s	remaining: 1h 7m 18s
900:	learn: 0.6791433	test: 0.6940283	best: 0.6944856 (807)	total: 27m 48s	remaining: 1h 4m 47s
Stopped by overfitting

In [45]:
print(scores)
print(np.mean(scores))

[0.6944856461576013, 0.6940264307292299, 0.6910995745637872, 0.6903847234975544, 0.6895593824785322]
0.6919111514853411


l포함+k3일때
Learning rate set to 0.086395
0:	learn: 0.6325044	test: 0.6361368	best: 0.6361368 (0)	total: 1.34s	remaining: 1h 6m 52s
100:	learn: 0.6573345	test: 0.6864060	best: 0.6864060 (100)	total: 2m 43s	remaining: 1h 18m 23s
200:	learn: 0.6646478	test: 0.6912543	best: 0.6914050 (198)	total: 5m 27s	remaining: 1h 16m
300:	learn: 0.6683587	test: 0.6927556	best: 0.6928172 (293)	total: 8m 12s	remaining: 1h 13m 33s
400:	learn: 0.6710274	test: 0.6934692	best: 0.6935095 (399)	total: 11m	remaining: 1h 11m 21s
500:	learn: 0.6729956	test: 0.6937643	best: 0.6938669 (487)	total: 13m 54s	remaining: 1h 9m 20s
600:	learn: 0.6748262	test: 0.6940238	best: 0.6941299 (553)	total: 16m 49s	remaining: 1h 7m 8s
700:	learn: 0.6766024	test: 0.6941486	best: 0.6943747 (614)	total: 20m	remaining: 1h 5m 35s
Stopped by overfitting detector  (100 iterations wait)

bestTest = 0.6943747425
bestIteration = 614

In [22]:
a = model.predict_proba(X_test)[:, 1]

In [24]:
b = np.where(a > threshold , 1, 0)

In [ ]:
#6.79

In [19]:
threshold = 0.4
pred_list = []
scores = []
for i,(tri, vai) in enumerate( cv.split(X_train) ):
    pred = models[i].predict_proba(X_train.iloc[vai])[:, 1]
    pred = np.where(pred >= threshold , 1, 0)
    score = f1_score(y_train[vai],pred)
    scores.append(score)
    pred = models[i].predict_proba(X_test)[:, 1]
    pred_list.append(pred)
print(scores)
print(np.mean(scores))

NameError: name 'models' is not defined

In [53]:
pred = np.mean( pred_list , axis = 0 )
pred = np.where(pred >= threshold , 1, 0)

In [25]:
submission = pd.read_csv('./_data/sample_submission.csv')
submission.head()

id  target
0   0       0
1   1       0
2   2       0
3   3       0
4   4       0

In [26]:
submission['target'] = b
submission

id  target
0          0       0
1          1       0
2          2       0
3          3       0
4          4       1
...      ...     ...
46399  46399       1
46400  46400       1
46401  46401       1
46402  46402       1
46403  46403       1

[46404 rows x 2 columns]

In [27]:
submission.to_csv('./submission15.csv', index=False)

# public 0.70096